addons/account_edi/models/account_edi_format.py

```diff
@@ -506,21 +506,90 @@ def _retrieve_partner(self, name=None, phone=None, mail=None, vat=None):
         :param vat:     The vat number of the partner.
         :returns:       A partner or an empty recordset if not found.
         '''
-        domains = []
-        for value, domain in (
-            (name, [('name', 'ilike', name)]),
-            (phone, expression.OR([[('phone', '=', phone)], [('mobile', '=', phone)]])),
-            (mail, [('email', '=', mail)]),
-            (vat, [('vat', 'like', vat)]),
-        ):
-            if value is not None:
-                domains.append(domain)
-
-        domain = expression.AND([
-            expression.OR(domains),
-            [('company_id', 'in', [False, self.env.company.id])],
-        ])
-        return self.env['res.partner'].search(domain, limit=1)
+        def search_with_vat(extra_domain):
+            if not vat:
+                return None
+
+            # Sometimes, the vat is specified with some whitespaces.
+            normalized_vat = vat.replace(' ', '')
+
+            partner = self.env['res.partner'].search([('vat', '=', normalized_vat)], limit=1)
+
+            # Try to remove the country code prefix from the vat.
+            if not partner and len(normalized_vat) > 2 and normalized_vat[:2].isalpha():
+                country_prefix = normalized_vat[:2]
+                normalized_vat = normalized_vat[2:]
+                partner = self.env['res.partner'].search(extra_domain + [
+                    ('vat', '=', normalized_vat),
+                    ('country_id.code', '=', country_prefix.lower()),
+                ], limit=1)
+
+                # The country could be not specified.
+                if not partner:
+                    partner = self.env['res.partner'].search(extra_domain + [
+                        ('vat', '=', normalized_vat),
+                        ('country_id', '=', False),
+                    ], limit=1)
+
+            # The vat could be a string of alphanumeric values without country code but with missing zeros at the
+            # beginning.
+            if not partner:
+                try:
+                    vat_only_alphanumeric = str(int(normalized_vat))
+                except ValueError:
+                    vat_only_alphanumeric = None
+
+                if vat_only_alphanumeric:
+                    query = self.env['res.partner']._where_calc([('active', '=', True)], extra_domain)
```

Comment:
`+` instead of `,`
You could probably use `expression.AND` too

Reply:
Funny thing is that the second parameter is `active_test`, which default to `True`, and is adding the `[('active', '=', True)]` domain without you needing to do it.